be/src/olap/rowset/segment_v2/row_ranges.h

```diff
@@ -0,0 +1,332 @@
+    }
+
+private:
+    int64_t _from;
+    int64_t _to;
+};
+
+class RowRanges {
+public:
+    RowRanges() : _cur_range_id(0) { }
+
+    RowRanges(RowRanges&& other) {
+        _ranges = std::move(other._ranges);
+        _cur_range_id = other._cur_range_id;
+    }
+
+    RowRanges& operator=(RowRanges&& other) {
+        if (this != &other) {
+            _ranges = std::move(other._ranges);
+            _cur_range_id = other._cur_range_id;
+        }
+        return *this;
+    }
+
+    /*
+    * Creates a new RowRanges object with the single range [0, row_count - 1].
+    */
+    static RowRanges create_single(uint64_t row_count) {
+        RowRanges ranges;
+        ranges.add(Range(0, row_count - 1));
+        return ranges;
+    }
+
+    /*
+    * Creates a new RowRanges object with the single range [from, to].
+    */
+    static RowRanges create_single(uint64_t from, uint64_t to) {
+        DCHECK(from <= to);
+        RowRanges ranges;
+        ranges.add(Range(from, to));
+        return ranges;
+    }
+
+    /*
+    * Calculates the union of the two specified RowRanges object. The union of two range is calculated if there are
+    * elements between them. Otherwise, the two disjunct ranges are stored separately.
+    * For example:
+    * [113, 241] ∪ [221, 340] = [113, 330]
+    * [113, 230] ∪ [231, 340] = [113, 340]
+    * while
+    * [113, 230] ∪ [232, 340] = [113, 230], [232, 340]
+    *
+    */
+    static void ranges_union(const RowRanges& left, const RowRanges& right, RowRanges* result) {
+        RowRanges tmp_range;
+        auto it1 = left._ranges.begin();
+        auto it2 = right._ranges.begin();
+        // merge and add
+        while (it1 != left._ranges.end() && it2 != right._ranges.end()) {
+            if (it1->is_after(*it2)) {
+                tmp_range.add(*it2);
+                ++it2;
+            } else {
+                tmp_range.add(*it1);
+                ++it1;
+            }
+        }
+        while (it1 != left._ranges.end()) {
+            tmp_range.add(*it1);
+            ++it1;
+        }
+        while (it2 != right._ranges.end()) {
+            tmp_range.add(*it2);
+            ++it2;
+        }
+        *result = std::move(tmp_range);
+    }
+
+    /*
+    * Calculates the intersection of the two specified RowRanges object. Two ranges intersect if they have common
+    * elements otherwise the result is empty.
+    * For example:
+    * [113, 241] ∩ [221, 340] = [221, 241]
+    * while
+    * [113, 230] ∩ [231, 340] = <EMPTY>
+    *
+    * The result RowRanges object will contain all the row indexes there were contained in both of the specified objects
+    */
+    static void ranges_intersection(const RowRanges& left, const RowRanges& right, RowRanges* result) {
+        RowRanges tmp_range;
+        for (auto it1 = left._ranges.begin(); it1 != left._ranges.end(); ++it1) {
+            const Range& range1 = *it1;
+            int right_index = 0;
```

Comment:
what does right_index mean? It is 0 for every it1

Reply:
it it for optimization. there is a bug here. I will fix it.